include/odp/api/spec/comp.h

```diff
@@ -0,0 +1,873 @@
+	 *  compression algo given as comp_alg. Implementation should not
+	 *  support hash only operation on data. output should always contain
+	 *  data + hash.
+	 *
+	 */
+	odp_comp_hash_alg_t hash_algo;
+
+	/** parameters specific to compression */
+	odp_comp_alg_param_t algo_param;
+
+	/** Async mode completion event queue
+	 *
+	 * When mode = ODP_COMP_ASYNC, user should wait on ODP_EVENT_PACKET
+	 * with subtype ODP_EVENT_PACKET_COMP on this queue.
+	 *
+	 * By default, implementation enques completion events in-order-of
+	 * request submission and thus queue is considered ordered.
+	 *
+	 * Please note, behavior could be changed or enhanced
+	 * to queue event in-order-of their completion to enable
+	 * performance-oriented application to leverage hw offered parallelism.
+	 * However, this will be subject to application requirement and more
+	 * explicit defined use-case.
+	 *
+	 */
+	odp_queue_t compl_queue;
+} odp_comp_session_param_t;
+
+/**
+ * Comp API per packet operation result
+ *
+ */
+typedef struct odp_comp_packet_op_result_t {
+	/** Operation Return Code */
+	odp_comp_err_t err;
+
+	/** Output packet data range. Where,
+	 *
+	 *  offset = base offset within packet, current data is written at.
+	 *
+	 *  length = length of data written.
+	 *
+	 */
+	odp_packet_data_range_t out_data_range;
+} odp_comp_packet_op_result_t;
+
+/**
+ * Comp packet API per packet operation parameters
+ */
+typedef struct odp_comp_packet_op_param_t {
+	/** Session handle from creation */
+	odp_comp_session_t session;
+
+	/** Boolean indicating if current input is last chunk of data, where
+	 *
+	 *   true : last chunk
+	 *
+	 *   false: more to follow
+	 *
+	 * If set to true, indicates this is the last chunk of
+	 * data. After processing of last chunk of data is complete i.e.
+	 * call returned with any error code except ODP_COMP_ERR_OUT_OF_SPACE,
+	 * implementation should move algorithm to stateless mode
+	 * for next of batch of operation i.e. reset history,
+	 * insert 'End of Block' marker into compressed data stream(if
+	 * supported by algo).See deflate/zlib for interpretation of
+	 * stateless/stateful.
+	 *
+	 * For stateless compressions (ex ipcomp), last should be set to 'true'
+	 * for every input packet processing call.
+	 *
+	 * For compression + hash, digest will be available after
+	 * last chunk is processed completely. In case of
+	 * ODP_COMP_ERR_OUT_OF_SPACE, application should keep on calling
+	 * odp_comp_xxx() API with more output buffer unless call returns
+	 * with ODP_COMP_ERR_NONE or other failure code except
+	 *  ODP_COMP_ERR_OUT_OF_SPACE.
+	 */
+	odp_bool_t last;
+
+	/** Input data range to process */
+	odp_packet_data_range_t in_data_range;
+
+	/** Output packet data range.
+	 * Indicates where processed packet will be written
+	 * where length specifies,
+	 * length of available packet buffer at input, and
+	 * length of data written at output
+	 */
+	odp_packet_data_range_t out_data_range;
```

Comment:
"where length specifies,
length of available packet buffer at input, and
length of data written at output"

This specification should be clarified. "Input" refers refers to out_data_range.length as operation input param and  "output" refers to the same as operation output parameter.

In addition to length, offset may be updated at operation output (at least when new packet was allocated). That should be mentioned also.

Reply:
offset is not updated by spec as compression does not support auto-allocation of packet and starts data write at given offset only. For description, I rephrased it as below. Please approve.
 "	/** Output packet data range.
	 *
	 * out_data_range.offset = offset in output packet data buffer to start
	 * write at,
	 *
	 * out_data_range.length = length of available packet data buffer
	 * from given offset.
	 *
	 */

Reply:
"where length specifies,
length of available packet buffer at input, and
length of data written at output"

This needs to be more precise on what input/output refers to: " ... application sets .length to max number of bytes that the operation may output ... On success, the operations sets .length to number of bytes written into the packet ..."

I think auto-allocation would be the easiest way to use the API. Why that would not be supported ? Implementation has better change (and higher performance) to allocate enough packet memory for output packet. If not supported, there should be API defined way for application approximate large enough output packet size (in decompress).